{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting C:\\Anaconda\\Lib\\CAMP.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile C:\\Anaconda\\Lib\\CAMP.py\n",
    "    \n",
    "import datetime as dt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math as math \n",
    "import MathsUtilities as MUte\n",
    "from scipy.optimize import curve_fit\n",
    "\n",
    "CampConstants = {\n",
    "    'k':-0.17,\n",
    "    'VernalisationThreshold':1.0,\n",
    "    'MethalationThreshold':0.5,\n",
    "}\n",
    "\n",
    "def LARPTQ(PTQ,maxLAR,minLAR,PTQhf):\n",
    "    \"\"\"Calculate the Leaf Appearance Rate at a given PTQ for a genotype of given minLAR and MaxLAR as formualted by Baumont etal 2019 Journal Expt Botany, Equation 3. \n",
    "    Args:\n",
    "        PTQ: PhotoThermal Quatent (mmol PAR m-2 oCd-1)\n",
    "        maxLAR: Leaf Appearance Rate at PTQ = infinity (oCd-1)\n",
    "        minLAR: Leaf Appearance rate at PTQ = 0 (oCd-1)\n",
    "        PTQhf: PTQ half, controls the curvature of response\n",
    "    Returns:\n",
    "        Leaf Appearance Rate (dHS/oCd)\n",
    "    \"\"\"\n",
    "    return minLAR+((maxLAR-minLAR)*PTQ)/ (PTQhf+PTQ)\n",
    "\n",
    "def CalcPTQ(Tt,Pp,lightIntensity):\n",
    "    \"\"\"Calculate the Photothermal Quotient for a given day in controlled environment (mmol PAR m-2 oCd-1). \n",
    "    Args:\n",
    "        Tt: Mean daily thermal time (oC)\n",
    "        Pp: Photoperiod (oCd-1)\n",
    "        LightIntensity: radiation output when lights are on (MJ/m2/s)\n",
    "    Returns:\n",
    "        Photothermal Quotient (mmol PAR m-2 oCd-1)\n",
    "    \"\"\"\n",
    "    Radn = Pp * 3600 * lightIntensity * 2  # Hours * s/h * MJ/m2/s * Coeff_PAR->Total \n",
    "    Ft = 1\n",
    "    if Tt < 15:\n",
    "        Ft = Tt/15\n",
    "    return (Radn * 2.285 * Ft)/Tt\n",
    "\n",
    "def CalcdHS(Tt,HS,PTQ,maxLAR,minLAR,PTQhf):\n",
    "    \"\"\"Calculate daily increase in Haun stage\n",
    "    Args:\n",
    "        Tt: Thermal time increment\n",
    "        HS: Current Haun Stage\n",
    "        PTQ: PhotoThermalQuotient\n",
    "        maxLAR: Leaf Appearance Rate at PTQ = infinity (oCd-1)\n",
    "        minLAR: Leaf Appearance rate at PTQ = 0 (oCd-1)\n",
    "        PTQhf: PTQ half, controls the curvature of response\n",
    "    Returns:\n",
    "        Daily increas in Haun Stage\n",
    "    \"\"\"\n",
    "    BasePhyllochron = 1/LARPTQ(PTQ,maxLAR,minLAR,PTQhf)\n",
    "    StageFactor = np.interp(HS,[0,7,8,12],[1.0,1.0,1.4,1.4],False)\n",
    "    Phyllochron = BasePhyllochron*StageFactor\n",
    "    return Tt/Phyllochron\n",
    "\n",
    "def CalcPpResponse(Pp, Base, Max, dHS):\n",
    "    \"\"\" Calculate value of a photoperiod sensitive variable between Max and Baee values\n",
    "    Args:\n",
    "        Pp: Photoperiod\n",
    "        Base: value below 8h Pp\n",
    "        Max: value above 16h Pp\n",
    "        dHS: delta haun stage\n",
    "    Returns:\n",
    "        variable value dependent on photoperiod and adjusted for dHS\n",
    "    \"\"\"\n",
    "    if Pp <= 8.0:\n",
    "        return Base * dHS\n",
    "    if (Pp > 8.0) and (Pp < 16.0):\n",
    "        return (Base + (Max * (Pp-8)/(16-8))) * dHS\n",
    "    if (Pp >= 16.0):\n",
    "        return Max * dHS  \n",
    "\n",
    "def CalcBaseUpRegVrn1(Tt, dHS, BaseDVrn1):\n",
    "    \"\"\" Calculate upregulation of base Vrn1\n",
    "    \n",
    "    Args:\n",
    "        Tt: Thermal time increment\n",
    "        dHS: Delta Haun stage\n",
    "        BaseDVrn1: coeffociennt for Base Vrn1 expression\n",
    "        \n",
    "    Returns:\n",
    "        delta BaseVrn1 representing the additional Vrn1 expression from base expression\n",
    "    \"\"\"\n",
    "    if Tt < 0: \n",
    "        BaseDVrn1 = 0\n",
    "    return BaseDVrn1 * dHS\n",
    "\n",
    "def CalcColdUpRegVrn1(Tt,dHS, MaxDVrn1, k):\n",
    "    \"\"\" Upregulation of Vrn1 from cold.  Is additional to base vrn1\n",
    "        BaseDVrn1 in seperate calculation otherwise te same as Brown etal 2013\n",
    "    Args:\n",
    "        Tt: Thermal time increment\n",
    "        dHS: Delta Haun stage\n",
    "        MaxDVrn1: coefficient for Maximum upregulation of Vrn1\n",
    "        k: The exponential coefficient determining temperature response\n",
    "        \n",
    "    Returns:\n",
    "        delta ColdVrn1 representing the additional Vrn1 expression from cold upregulation\n",
    "    \"\"\"\n",
    "    UdVrn1 = MaxDVrn1 * np.exp(k*Tt)\n",
    "    if (Tt < 20):\n",
    "        return UdVrn1 * dHS\n",
    "    else:\n",
    "        return -5\n",
    "    \n",
    "def TSHS(FLN, IntFLNvsTSHS):\n",
    "    \"\"\"Haun stage timing of terminal spikelet \n",
    "       Inverts equation 5 from Brown etal 2013 FLN =  2.85 + 1.1*TSHS\n",
    "       Note the intercept differs, was typeo on publication\n",
    "       Intercept has been made variable for as needs to be lower for some very fast varieties\n",
    "    Args:\n",
    "        FLN: The final leaf number\n",
    "        IntFLNvsTSHS: The intercept of the regression between FLN and TSHS\n",
    "    Returns:\n",
    "        Estimation of the haun stage timing of terminal spikelet\n",
    "    \"\"\"\n",
    "    return (FLN - IntFLNvsTSHS)/1.1\n",
    "\n",
    "def CAMPmodel(Out, Day, Tt, Pp, Params, Consts, TtEmerge, LightIntensity):\n",
    "    \"\"\" The Cereal Anthesis Molecular Phenology model.\n",
    "        Based on the ideas presented in Brown etal 2014 (Annals of Botany)\n",
    "        Alterations made replacing Vrn4 notion with methalation of Vrn1, introducing VrnX to account for long day vernalisation\n",
    "        Vrn2 approach changed to have hihger experssion for first HS then a lower rate\n",
    "        Other alterations to implement working code base\n",
    "    Args:\n",
    "        Out: \"FLN\" or None.  If \"FLN\" will only return estimated FLN else will return full dataframe with daily state variable values\n",
    "        Day: List, 1:EndDay representing the timesteps in model run\n",
    "        Tt: List of same length as Day, representing daily temperature\n",
    "        Pp: List of same length as Day, representing daily photoperiod\n",
    "        Params: Dict with genotype parameters fitted by CalcCultivarVrnCoeffs() \n",
    "        Consts: Dict with crop specific constants,\n",
    "        TtEmerge: Thermaltime from sowing to emergence\n",
    "        LightIntensity: radiation intensity (MJ/m2/s)\n",
    "    \"\"\"\n",
    "    # Set up Data structure and initialise values\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    DF = pd.DataFrame(index = [0], columns = ['Day','Tt','Pp'])\n",
    "    IsGerminated = False\n",
    "    IsEmerged = False\n",
    "    IsDayOfEmergence = False\n",
    "    IsMethalating = False\n",
    "    IsVernalised = False\n",
    "    IsInduced = False\n",
    "    IsReproductive = False\n",
    "    IsAtFlagLeaf = False\n",
    "    DF.loc[0,'Day'] = 0\n",
    "    DF.loc[0,'Tt'] = Tt[0]\n",
    "    DF.loc[0,'Pp'] = Pp[0]\n",
    "    DF.loc[0,'PTQ'] = 0\n",
    "    DF.loc[0,'Stage'] = \"\"\n",
    "    DF.loc[0,'HS'] = 0.0\n",
    "    DF.loc[0,'AccumTt'] = 0\n",
    "    DF.loc[0,'BaseVrn1'] = 0\n",
    "    DF.loc[0,'ColdVrn1'] = 0\n",
    "    DF.loc[0,'MethColdVrn1'] = 0\n",
    "    DF.loc[0,'pVrn2'] = 0\n",
    "    DF.loc[0,'Vrn1'] = 0\n",
    "    DF.loc[0,'Vrn2'] = 0.0\n",
    "    DF.loc[0,'Vrn3'] = 0.0\n",
    "    DF.loc[0,'VrnX'] = 0.0\n",
    "    DF.loc[0,'VSHS'] = 0\n",
    "    DF.loc[0,'FIHS'] = 0\n",
    "    DF.loc[0,'TSHS'] = 0\n",
    "    DF.loc[0,'FLN'] = Params.IntFLNvsTSHS\n",
    "    DF.loc[0,'dHS'] = 0\n",
    "    DF.loc[0,'dBaseVrn1'] = 0\n",
    "    DF.loc[0,'dColdVrn1'] = 0\n",
    "    DF.loc[0,'dMethColdVrn1'] = 0\n",
    "    DF.loc[0,'dpVrn2'] = 0\n",
    "    DF.loc[0,'dVrn3'] = 0\n",
    "    DF.loc[0,'dVrnX'] = 0\n",
    "    DF.loc[0,'HSEmerge'] = 0\n",
    "    d = 1\n",
    "    BaseVrn1AtEmergence = 0\n",
    "    # Model daily loop\n",
    "    # ^^^^^^^^^^^^^^^^\n",
    "    while (IsAtFlagLeaf == False) and (d < Day[-1]):\n",
    "        #Sow crop on day 1\n",
    "        DF.loc[d,'Day'] = d\n",
    "        if (IsGerminated == False):\n",
    "            DF.loc[d,'Stage'] = 'Germination'\n",
    "            IsGerminated = True\n",
    "        # Set daily environment variables\n",
    "        DF.loc[d,'Tt'] = Tt[d-1]\n",
    "        DF.loc[d,'Pp'] = Pp[d-1]\n",
    "        #Zero set deltas\n",
    "        DF.loc[d,'dBaseVrn1'] = 0\n",
    "        DF.loc[d,'dColdVrn1'] = 0\n",
    "        DF.loc[d,'dMethColdVrn1'] = 0\n",
    "        DF.loc[d,'dpVrn2'] = 0\n",
    "        DF.loc[d,'dVrn3'] = 0\n",
    "        DF.loc[d,'dVrnX'] = 0\n",
    "                \n",
    "        DF.loc[d,'AccumTt'] = DF.loc[d-1,'AccumTt'] + DF.loc[d,'Tt']\n",
    "        \n",
    "        #PropnOfDay = 1.0\n",
    "        if (DF.loc[d,'AccumTt'] > TtEmerge) and (IsEmerged==False):\n",
    "            IsEmerged = True\n",
    "            IsDayOfEmergence = True\n",
    "            DF.loc[d,'Stage'] = 'Emergence'\n",
    "            #PropnOfDay = (DF.loc[d,'AccumTt'] - TtEmerge)/DF.loc[d,'Tt'] # Calculate fraction of emergence days Tt that is not used for emergence\n",
    "        \n",
    "    # Calculate daily Haun Stage changes\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "        DF.loc[d,'PTQ'] = CalcPTQ(DF.loc[d,'Tt'],DF.loc[d,'Pp'],LightIntensity)\n",
    "        if IsEmerged==True: # Crop emerged\n",
    "            # Calculate delta haun stage\n",
    "            DF.loc[d,'dHS'] = CalcdHS(DF.loc[d,'Tt'],DF.loc[d-1,'HS'],DF.loc[d,'PTQ'],Params.maxLAR,Params.minLAR,Params.PTQhf)\n",
    "        else: # Crop not yet emerged but we still need a dHS value to drive Vrn1 expression prior to emergence\n",
    "            EmergDurationFactor = 1\n",
    "            if (DF.loc[d,'AccumTt'] > 90):  # Calculate EmergenceDurationFactor to slow accumulation of HS if emergence is taking a long time.  This slows Vrn1 expression under slow emergence and strange responses to delayed sowing\n",
    "                EmergDurationFactor = 1 * np.exp(-0.015 * (DF.loc[d,'AccumTt']-90))\n",
    "            DF.loc[d,'dHS']  = DF.loc[d,'Tt']/(1/LARPTQ(1.0,Params.maxLAR,Params.minLAR,Params.PTQhf)) * EmergDurationFactor\n",
    "            DF.loc[d,'HSEmerge'] = DF.loc[d-1,'HSEmerge'] + DF.loc[d,'dHS'] #* PropnOfDay\n",
    "       \n",
    "        #increment HS\n",
    "        if IsEmerged == True:\n",
    "            DF.loc[d,'HS'] = DF.loc[d-1,'HS'] +  DF.loc[d,'dHS']\n",
    "        else:\n",
    "            DF.loc[d,'HS'] = 0\n",
    "    \n",
    "    # Calculate Vrn gene expression\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "        if (IsVernalised==False): # do vernalisation calculations if crop not yet vernalised\n",
    "        # Calculate Vrn1 deltas\n",
    "            DF.loc[d,'dBaseVrn1'] = CalcBaseUpRegVrn1(DF.loc[d,'Tt'],DF.loc[d,'dHS'],Params.BaseDVrn1) \n",
    "            DF.loc[d,'dColdVrn1'] = CalcColdUpRegVrn1(DF.loc[d,'Tt'],DF.loc[d,'dHS'],Params.MaxDVrn1,Consts['k'])\n",
    "            DF.loc[d,'ColdVrn1'] = max(0.0,DF.loc[d-1,'ColdVrn1'] + DF.loc[d,'dColdVrn1'])\n",
    "            \n",
    "        # Calculate daily methalation\n",
    "            if DF.loc[d,'ColdVrn1']>= Consts['MethalationThreshold']: # ColdVrn1 expressed to threshold required for methalation to occur\n",
    "                IsMethalating = True\n",
    "            else:\n",
    "                IsMethalating = False\n",
    "                \n",
    "            if (IsMethalating == True):\n",
    "                DF.loc[d,'dMethColdVrn1'] = min(DF.loc[d,'ColdVrn1'] - Consts['MethalationThreshold'],\n",
    "                                                max(0,DF.loc[d,'dColdVrn1']))\n",
    "            \n",
    "        # Calcualte expression of photoperiod sensitive genes\n",
    "            DF.loc[d,'dVrnX'] = 0.0\n",
    "            DF.loc[d,'dpVrn2'] = 0.0\n",
    "            if (IsEmerged==True):  # Photoperiod sensitive genes only express after emergence\n",
    "                if DF.loc[d-1,'MethColdVrn1'] == 0.0:  # VrnX expression only occurs if no methalation of Vrn1 has occured\n",
    "                    DF.loc[d,'dVrnX'] = CalcPpResponse(DF.loc[d,'Pp'], 0.0, Params.MaxDVrnX, DF.loc[d,'dHS'])\n",
    "                \n",
    "                if IsDayOfEmergence == True: # Calculate high dpVrn2 on emergence day to give initial value\n",
    "                    BaseVrn1AtEmergence = DF.loc[d-1,'BaseVrn1'] + DF.loc[d,'dBaseVrn1']\n",
    "                    DF.loc[d,'dpVrn2'] = CalcPpResponse(DF.loc[d,'Pp'],BaseVrn1AtEmergence,Params.MaxIpVrn2,1.0)\n",
    "                else:\n",
    "                    DF.loc[d,'dpVrn2'] = CalcPpResponse(DF.loc[d,'Pp'],0,Params.MaxDpVrn2,DF.loc[d,'dHS'])\n",
    "                    \n",
    "                \n",
    "        # Increment todays Vrn expression values using deltas just calculated\n",
    "            DF.loc[d,'pVrn2'] = DF.loc[d-1,'pVrn2'] + DF.loc[d,'dpVrn2']\n",
    "            DF.loc[d,'VrnX'] = DF.loc[d-1,'VrnX'] + DF.loc[d,'dVrnX']\n",
    "            DF.loc[d,'BaseVrn1'] = DF.loc[d-1,'BaseVrn1'] + DF.loc[d,'dBaseVrn1']\n",
    "            DF.loc[d,'MethColdVrn1'] = DF.loc[d-1,'MethColdVrn1'] + DF.loc[d,'dMethColdVrn1'] \n",
    "            # Effective expression of Vrn1 is the sum of baseVrn1, MethalatedVrn1 and Vrnx expression\n",
    "            DF.loc[d,'Vrn1'] = DF.loc[d-1,'Vrn1'] + DF.loc[d,'dMethColdVrn1'] \\\n",
    "                                                          + DF.loc[d,'dBaseVrn1'] \\\n",
    "                                                          + DF.loc[d,'dVrnX'] \n",
    "            # Effective Vrn2 expression is the potential expression less that which is blocked by Vrn1\n",
    "            DF.loc[d,'Vrn2'] =  max(0.0,DF.loc[d,'pVrn2'] - DF.loc[d,'Vrn1'])\n",
    "                                    \n",
    "        # Once Vernalisation complete, carry over vern expression values from yesterday\n",
    "        else:\n",
    "            DF.loc[d,'BaseVrn1'] = DF.loc[d-1,'BaseVrn1']\n",
    "            DF.loc[d,'ColdVrn1'] = DF.loc[d-1,'ColdVrn1'] \n",
    "            DF.loc[d,'MethColdVrn1'] = DF.loc[d-1,'MethColdVrn1'] \n",
    "            DF.loc[d,'Vrn1'] = DF.loc[d-1,'Vrn1']\n",
    "            DF.loc[d,'pVrn2'] = DF.loc[d-1,'pVrn2']\n",
    "            DF.loc[d,'Vrn2'] = DF.loc[d-1,'Vrn2']\n",
    "            DF.loc[d,'VrnX'] = DF.loc[d-1,'VrnX']\n",
    "            \n",
    "    # Set Haun stage variables\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "        if IsVernalised == False:\n",
    "            DF.loc[d,'VSHS'] = DF.loc[d,'HS']\n",
    "        else:\n",
    "            DF.loc[d,'VSHS'] = DF.loc[d-1,'VSHS']\n",
    "        if IsInduced == False:\n",
    "            DF.loc[d,'FIHS'] = DF.loc[d,'HS']\n",
    "        else: \n",
    "            DF.loc[d,'FIHS'] = DF.loc[d-1,'FIHS']\n",
    "        if IsReproductive == False:\n",
    "            DF.loc[d,'TSHS'] = DF.loc[d,'HS']\n",
    "            DF.loc[d,'FLN'] = Params.IntFLNvsTSHS + 1.1 * DF.loc[d,'TSHS']\n",
    "        else:\n",
    "            DF.loc[d,'TSHS'] = DF.loc[d-1,'TSHS']\n",
    "            DF.loc[d,'FLN'] = DF.loc[d-1,'FLN']\n",
    "\n",
    "    # Work out if vernalisation is complete    \n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "        # Vernalisation saturation occurs when Vrn1 > the vernalisation threshold and Vrn2 expression is zero\n",
    "        if (IsEmerged==True) and (DF.loc[d,'Vrn1'] >= Consts['VernalisationThreshold']) and (DF.loc[d,'Vrn2'] == 0) and (IsVernalised==False):\n",
    "            IsVernalised  = True\n",
    "            DF.loc[d,'Stage'] = 'Vern Sat'\n",
    "        \n",
    "    # Then work out Vrn3 expression\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "        # Vrn3 expression occurs between Vernalisation saturation and Terminal spikelet\n",
    "        if (IsVernalised == True) and (IsReproductive == False):\n",
    "            DF.loc[d,'dVrn3'] = CalcPpResponse(DF.loc[d,'Pp'], Params.BaseDVrn3, Params.MaxDVrn3, DF.loc[d,'dHS'])\n",
    "        DF.loc[d,'Vrn3'] = min(1,DF.loc[d-1,'Vrn3'] + DF.loc[d,'dVrn3'])\n",
    "\n",
    "    # Finally determine phenological stage\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "        # Then work out phase progression based on Vrn expression\n",
    "        if (DF.loc[d,'Vrn3'] >= 0.3) and (IsInduced == False):\n",
    "            IsInduced = True;\n",
    "        if (DF.loc[d,'Vrn3'] >=  1.0) and (IsReproductive == False):\n",
    "            IsReproductive = True;\n",
    "            DF.loc[d,'Stage'] = 'Term Spike'\n",
    "        \n",
    "        #Work out if Flag leaf has appeared.\n",
    "        if DF.loc[d,'HS'] >= DF.loc[d,'FLN']:\n",
    "            IsAtFlagLeaf = True    \n",
    "            DF.loc[d,'Stage'] = 'FlagLeaf'\n",
    "    \n",
    "        #Add states to dataframe\n",
    "        IsDayOfEmergence = False\n",
    "        \n",
    "        DF.loc[d,'IsGerminated'] = IsGerminated\n",
    "        DF.loc[d,'IsEmerged'] = IsEmerged\n",
    "        DF.loc[d,'IsMethalating'] = IsMethalating\n",
    "        DF.loc[d,'IsVernalised'] = IsVernalised\n",
    "        DF.loc[d,'IsReproductive'] = IsReproductive\n",
    "        DF.loc[d,'IsAtFlagLeaf'] = IsAtFlagLeaf\n",
    "    \n",
    "        # Increment day\n",
    "        d += 1\n",
    "        \n",
    "    if (Out == 'FLN') and (IsAtFlagLeaf == True):\n",
    "        return DF.iloc[-1,:]['FLN']\n",
    "    else:\n",
    "        return DF\n",
    "    \n",
    "def CalcCultivarVrnCoeffs(inputs):\n",
    "    \"\"\" Calculate VrnRate paremeters for genotype from given FLN observations under set vernalisation and photoperiod treatments\n",
    "    Args:\n",
    "        inputs: Dict or Series containing the following.\n",
    "        FLN_LV: Final Leaf Number under long day full vernalisation treatment\n",
    "        FLN_LN: Final Leaf Number under long day un-vernalisation treatment\n",
    "        FLN_SV: Final Leaf Number under short day full vernalisation treatment\n",
    "        FLN_SN: Final Leaf Number under short day un-vernalisation treatment\n",
    "        VrnTreatTemp: Vernalisation Treatment Temperature\n",
    "        VrnTreatDuration: Days exposure to vernalisting temperature\n",
    "        TreatmentPTQ_L: The PTQ under long Pp,\n",
    "        TreatmentPTQ_S: The PTQ under short Pp,\n",
    "        TtEmerge: Observed Thermal time from sowing to emergence\n",
    "        k: Vernalisation temperature response coefficient\n",
    "        VernSatThreshold: The amount of Methalated Vrn1 that is required for vernilisation saturation to occur\n",
    "        MethVrn1Threshold: The amount of cold Vrn1 expression required for methalation of Vrn1 to start\n",
    "        maxLAR: Leaf Appearance Rate at PTQ = infinity (oCd-1)\n",
    "        minLAR: Leaf Appearance rate at PTQ = 0 (oCd-1)\n",
    "        PTQhf: PTQ half, controls the curvature of response\n",
    "    Returns:\n",
    "        Series with values assigned to each variable in CultivarParameterList \n",
    "    \"\"\"\n",
    "    #Set up Data Store\n",
    "    data = pd.Series()\n",
    "    \n",
    "    # copy some params into outputs for checking\n",
    "    data['maxLAR'] = inputs['maxLAR']\n",
    "    data['minLAR'] = inputs['minLAR']\n",
    "    data['PTQhf']  = inputs['PTQhf']\n",
    "    data['FLN_LV'] = inputs['FLN_LV']\n",
    "    data['FLN_LN'] = inputs['FLN_LN']\n",
    "    data['FLN_SV'] = inputs['FLN_SV']\n",
    "    data['FLN_SN'] = inputs['FLN_SN']\n",
    "    \n",
    "###########################################\n",
    "# Calculate phase durations (in Haun Stage)\n",
    "###########################################\n",
    "    \n",
    "    # Haun stage duration of Emergence (EmergHS), assume a PTQ of 1.0 suitable for before emergence\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['EmergHS'] = inputs['TtEmerge']/(1/LARPTQ(1.0,inputs['maxLAR'],inputs['minLAR'],inputs['PTQhf']))\n",
    "    \n",
    "    # Haun stage duration of vernalisation treatment period\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    # If Vern treatment complete before emergence (VernTreatHS_L,VernTreatHS_S), assume a PTQ of 1.0 suitable for before emergence\n",
    "    data['VrnTreatTtDurat'] = inputs['VrnTreatDuration'] * inputs['VrnTreatTemp']\n",
    "    if data['VrnTreatTtDurat'] <= inputs['TtEmerge']: \n",
    "        data['VernTreatHS_L'] = data['VrnTreatTtDurat']/\\\n",
    "                                (1/LARPTQ(1.0,inputs['maxLAR'],inputs['minLAR'],inputs['PTQhf']))\n",
    "        data['VernTreatHS_S'] = data['VrnTreatTtDurat']/\\\n",
    "                                (1/LARPTQ(1.0,inputs['maxLAR'],inputs['minLAR'],inputs['PTQhf']))\n",
    "    # If Vern treatment completes after emergence, add the HS when treatment finished to Emergence HS\n",
    "    else: \n",
    "        data['EmergToEndTreatHS_L'] = (data['VrnTreatTtDurat'] - inputs['TtEmerge'])/\\\n",
    "                                      (1/LARPTQ(inputs['TreatmentPTQ_L'],inputs['maxLAR'],inputs['minLAR'],inputs['PTQhf'])) \n",
    "        data['EmergToEndTreatHS_S'] = (data['VrnTreatTtDurat'] - inputs['TtEmerge'])/\\\n",
    "                                      (1/LARPTQ(inputs['TreatmentPTQ_S'],inputs['maxLAR'],inputs['minLAR'],inputs['PTQhf'])) \n",
    "        data['VernTreatHS_L'] = data['EmergHS'] + data['EmergToEndTreatHS_L']\n",
    "        data['VernTreatHS_S'] = data['EmergHS'] + data['EmergToEndTreatHS_S']\n",
    "    \n",
    "    # The soonest a plant may exhibit vern saturation if vernalisation completed early\n",
    "    MinVSHS = 1.1\n",
    "    \n",
    "    # Minimum Haun stage duration from vernalisation saturation to terminal spikelet under long day conditions (MinHSVsTs)\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    # Assume maximum of 3, Data from Lincoln CE (CRWT153) showed varieties that harve a high TSHS hit VS ~3HS prior to TS \n",
    "    data['MinHSVsTs'] = min(3,(data['FLN_LV'] - MinVSHS)/2)\n",
    "    MinVSHS_FLN_LV = data['FLN_LV'] - MinVSHS - data['MinHSVsTs']\n",
    "    data['IntFLNvsTSHS'] = min(2.85,MinVSHS_FLN_LV/1.1)\n",
    "    \n",
    "    # Calculate Terminal spikelet duration (TSHS) for each treatment from FLNData\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['TSHS_LV'] = TSHS( data['FLN_LV'], data['IntFLNvsTSHS'])\n",
    "    data['TSHS_LN'] = TSHS( data['FLN_LN'], data['IntFLNvsTSHS'])\n",
    "    data['TSHS_SV'] = TSHS( data['FLN_SV'], data['IntFLNvsTSHS'])\n",
    "    data['TSHS_SN'] = TSHS( data['FLN_SN'], data['IntFLNvsTSHS'])\n",
    "        \n",
    "    # Photoperiod sensitivity (PPS) is the difference between TSHS at 8 and 16 h pp under full vernalisation.\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['PPS'] = max(0,data['TSHS_SV'] - data['TSHS_LV'])\n",
    "    \n",
    "    # Vernalisation Saturation duration (VSHS) for each environment from TSHS and photoperiod response\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    # Terminal spikelet duration less the minimum duration from VS to TS under long day treatment\n",
    "    data['VSHS_LV'] = max(MinVSHS, data['TSHS_LV'] - data['MinHSVsTs']) \n",
    "    data['VSHS_LN'] = max(MinVSHS, data['TSHS_LN'] - data['MinHSVsTs']) \n",
    "    # Terminal spikelet duration less the minimum duration from VS to TS and the photoperiod extension of VS to TS under short day treatment\n",
    "    data['VSHS_SV'] = max(MinVSHS, data['TSHS_SV'] - (data['MinHSVsTs'] + data['PPS']))\n",
    "    data['VSHS_SN'] = max(MinVSHS, data['TSHS_SN'] - (data['MinHSVsTs'] + data['PPS']))\n",
    "    \n",
    "####################################\n",
    "# Calculate Photoperiod sensitivities\n",
    "####################################\n",
    "    \n",
    "    # Maximum delta for Upregulation of Vrn3 (MaxDVrn3)\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    # Occurs under long Pp conditions, Assuming Vrn3 increases from 0 - VernSatThreshold between VS to TS and this takes \n",
    "    # MinHSVsTs under long Pp conditions.\n",
    "    data['MaxDVrn3'] = inputs['VernalisationThreshold'] / data['MinHSVsTs']\n",
    "    \n",
    "        \n",
    "    # Base delta for upredulation of Vrn3 (BaseDVrn3)\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    # Occurs under short Pp conditions, Assuming Vrn3 increases from 0 - VernSatThreshold from VS to TS \n",
    "    # and this take 3 HS plus the additional HS from short Pp delay.\n",
    "    data['BaseDVrn3'] = inputs['VernalisationThreshold'] / (data['MinHSVsTs'] + data['PPS'])\n",
    "    \n",
    "    ## Under long day conditions Vrn3 expresses at its maximum rate and plant moves quickley from VS to TS.\n",
    "    ## Genotypic variation in MaxDVrn3 will contribute to differences in earlyness per se\n",
    "    ## Under shord day condition Vrn3 expressssion may be slower, taking longer to get from VS to TS\n",
    "    ## Photoperiod sensitiviey of a genotype is determined by differences in BaseVrn3 and MaxVrn3.  \n",
    "    ## if the two values are the same the genotype will not show photoperiod sensitivity\n",
    "    ## the greater the difference the more resonse the genotype will show to photoperiod\n",
    "    \n",
    "#################################\n",
    "# Calculate Base development rate \n",
    "#################################\n",
    "\n",
    "    # Base delta for upregulation of Vrn1 (BaseDVrn1) \n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    # The rate of expression is calculated as the amount of expression divided by the duration\n",
    "    # Under short Pp treatment Vrn2 is absent so the amount of Vrn1 expression required for VS to occur is given by VernalisationThreshold\n",
    "    # Under non vernalising conditions Vrn1 expression will happen at the base rate.  \n",
    "    # VrnX expression is absent in short days so baseVrn1 is the only contributor to the timing of VS\n",
    "    # BaseVrn1 expression starts when the seed is imbibed so the duration of expression is emergence plus VS\n",
    "    data['BaseDVrn1'] = inputs['VernalisationThreshold'] / (data['VSHS_SN'] + data['EmergHS'])\n",
    "    \n",
    "    # Genotypic variation in BaseDVrn1 contributes to intrinsic earlyness.\n",
    "    # A genotype with high BaseDVrn1 will reach VS quickly regardless of vernalisation exposure\n",
    "    # A genotype with low BaseDVrn1 will reach VS slowly if not vernalised but the duration of VS may be decreased\n",
    "    # by exposure to vernalising temperatures depending on cold vernalisation sensitivity\n",
    "    \n",
    "#############################################\n",
    "# Cold Vernalisation Sensitivity Calculations\n",
    "#############################################\n",
    "    \n",
    "    # Calcualtions of vernalisation sensitivity use data from short Pp treatments because Vrn2 and Vrnx are absent\n",
    "    # in these conditions and measured vernalistion response will be the result of Vrn1 expression alone.\n",
    "    \n",
    "    # Vernalisation Sensitivity (VS) measured simply as the difference between SV and SN treatuemnts\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['VS'] = data['VSHS_SN'] - data['VSHS_SV'] \n",
    "    \n",
    "    # To determine the effects of vernalisation treatment on Vrn1 expression we need to seperate these from baseVrn1 expression\n",
    "    # BaseVrn1 expression at VSHS ('BaseVrn1AtVSHS_SV')\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['BaseVrn1AtVSHS_SV'] = (data['VSHS_SV']+data['EmergHS']) * data['BaseDVrn1']\n",
    "    \n",
    "    # Any accelleration in VS due to cold exposure under short Photoperiod will be due to methatalated Vrn1 expression\n",
    "    # The amount of Vrn1 required for VS is given by the Vernalisation Threshold so the amount of methalated Vrn expression\n",
    "    # at VS must be this threshold less the amount of vrn1 contributed by base Vrn1 expression over this duration\n",
    "    # Methalated Cold Vrn1 expression under short Pp vernalisiation treatment (MethColdVern1AtTrans_SV)\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['MethColdVern1AtTrans_SV'] = max(0,inputs['VernalisationThreshold'] - data['BaseVrn1AtVSHS_SV'])\n",
    "    \n",
    "    # Cold Vrn 1 expression must first be upregulated to a methalation threshold and any further expression beyond this \n",
    "    # threshold is methalated into a persistant vernalisation response.  Thus the total expression of cold Vrn1 due to\n",
    "    # the vernalisation treatment applied is givin by MethColdVern1AtTrans_SV plus the Methalation Threshold\n",
    "    # Cold Vern1 expression at VSHS under short Pp vernalised conditions (ColdVrn1AtVSHS_SV)\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['ColdVrn1AtVSHS_SV'] = inputs['MethalationThreshold'] + data['MethColdVern1AtTrans_SV']\n",
    "    \n",
    "    # The haun stage duration over which vernalisation temperatures will have an effect is the minimum of the cold \n",
    "    # treatment duration (VernTreatHS) and the HS when vernalisation occurs as cold exposure after this is irrelevent\n",
    "    # Effective HS duration of cold treatment (EffectiveColdHS)\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['EffectiveColdHS'] = min(data['VSHS_SV']+data['EmergHS'], data['VernTreatHS_S']);\n",
    "    \n",
    "    # Then we calculate the rate as the amount divided by the duration\n",
    "    # Rate of cold Vrn1 expression at the vernalisation treatment temperture ('DVrn1AtVrnTreatTemp')\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['DVrn1AtVrnTreatTemp'] = data['ColdVrn1AtVSHS_SV']/data['EffectiveColdHS']\n",
    "    \n",
    "    # This rate is dependent on the temperature of the duration treatment and can be extrapolated to the maximum rate (at 0oC)\n",
    "    # using the exponential funciton proposed by Brown etal 2013 Annals of Botany.\n",
    "    # dVrn1 = MaxdVrn1 * np.exp(k*VrnTreatTemp) as DVrn1 At VrnTreatTemp' is known \n",
    "    # MaxDVrn1 is set to vero for genotypes with VS < 0.5 as these varieties are insensitive and calculated rates of MaxDVrn1 are simply amplifying noise\n",
    "    # Maximum upregulation delta Vrn1 (MaxDVrn1)\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    if data['VS']  > 0:\n",
    "        data['MaxDVrn1'] = data['DVrn1AtVrnTreatTemp']/ np.exp(inputs['k'] * inputs['VrnTreatTemp'])\n",
    "    else:\n",
    "        data['MaxDVrn1'] = 0.0\n",
    "    \n",
    "    # Genotypic variation in MaxDVrn1 combine with BaseDVrn1 to determine cold temperature vernalisatin sensitiviy\n",
    "    # A genotype with low BaseDVrn1 and high MaxDVrn1 will show high vernalisation sensitivity and sensitivity will decline \n",
    "    # either BaseDVrn1 increasees of MaxDVrn1 decreases.\n",
    "                                \n",
    "######################################\n",
    "# Photoperiod effects on vernalisation\n",
    "######################################\n",
    "    \n",
    "    # The parameters calculated above deal with photoperiod sensitivity in fully vernalised crops and vernalisation sensitivity\n",
    "    # under short photoperiod where cold is the only factor driving vernalisation response.\n",
    "    # Under long days photoperiod can also interact with vernalisation, either slowing the rate of vernalisation or speding it up\n",
    "    \n",
    "    # Vernalisation photoperiod sensitivy parameter (VPPS) is calculated to determine what effect photoperiod will have.\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['VPPS'] = 1-((data['VSHS_LN'] + data['EmergHS']) * data['BaseDVrn1'])\n",
    "    \n",
    "    # Genotypes that have a negative VPPS demonstrate short day vernalisation.  Under long days vernalisation requirement is reduced\n",
    "    # This is caused by the expression of Vrn2 which must be blocked by additional expression of Vrn1 meaning vernalisation takes longer.  \n",
    "    # The assumed mechanisum for Vrn2 is that it is expressed to a potential level and that one unit of Vrn1 will block the effective \n",
    "    # expression of Vrn2 so actual Vrn2 expression will always be lower than potential\n",
    "    # We can make some assumptions about the amounts of Vrn2 expression under long Pp conditions to calcualte rates                                \n",
    "    # Firstly, under un-vernalised conditions we can calculate the potential Vrn2 expression simply from\n",
    "    # the amount of BaseVrn1 expression up to VSHS\n",
    "                                \n",
    "    # Potential Vern2 expression at VS under long Pp UnVerrnalised treatment (pVrn2AtVSHS_LV)\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['pVrn2AtVSHS_LN'] = (data['VSHS_LN'] + data['EmergHS']) * data['BaseDVrn1']\n",
    "                                \n",
    "    # Under full vernalised treatment Vrn2 expression must be less than or equal to VernalisationThreshold at the time of VS.\n",
    "    # If we assume it is equal to VernalisationThreshold this provides us with another amount of pVrn2\n",
    "    # Potential Vern2 expression at VS under long Pp Vernalised treatment (pVrn2AtVSHS_LV)\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['pVrn2AtVSHS_LV'] = inputs['VernalisationThreshold']\n",
    "    \n",
    "    # If we regress these two potential Vrn2 amounts against their durations to VS the slope give us a rate of Vrn2 expression\n",
    "    # The resulting rate needs to always be less than BaseDVrn1 so Vrn1 can catch up with Vrn2 to cause vernalisation\n",
    "    # the amount of Vrn2 needs to quickly exceed BaseDVrn1 soon after emergence to achieve a delay in VS.\n",
    "    # The intercept of the regression quantifys how much Vrn2 would be expressed at emergence.  \n",
    "\n",
    "    # Initial potential Vrn2 at emergence (IPVrn2) and potential Vrn2 rate there after (DpVrn2)\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['MaxIpVrn2'] = 0\n",
    "    data['MaxDpVrn2'] = 0 \n",
    "    #Vrn2 parameters only relevent for genotypes with nevaitve VPPS and a low BaseDVrn1.  Genotypes with a high base\n",
    "    if (data['VPPS'] < 0): #and (data['BaseDVrn1'] < 0.4): \n",
    "        if data['VSHS_LN']-data['VSHS_LV'] == 0:\n",
    "            data['MaxIpVrn2'] = data['pVrn2AtVSHS_LN']\n",
    "        else:\n",
    "            data['MaxDpVrn2'] = max(0,(data['pVrn2AtVSHS_LN']-data['pVrn2AtVSHS_LV'])/(data['VSHS_LN']-data['VSHS_LV']))\n",
    "            data['MaxIpVrn2'] = (data['pVrn2AtVSHS_LV'])-(data['VSHS_LV']*data['MaxDpVrn2'])\n",
    "            if (data['MaxDpVrn2'] >= (data['BaseDVrn1']*0.99)): #If DpVrn2 exceeds baseVrn1 we need to do some forcing so it doesn't\n",
    "                data['BaseVrn1AtVSHS_LV'] = (data['VSHS_LV'] + data['EmergHS']) * data['BaseDVrn1']\n",
    "                data['MaxDpVrn2'] = max(0,(data['pVrn2AtVSHS_LN']-(data['BaseVrn1AtVSHS_LV']*1.2))/(data['VSHS_LN']-data['VSHS_LV']))\n",
    "                data['MaxIpVrn2'] = (data['pVrn2AtVSHS_LN'])-(data['VSHS_LN']*data['MaxDpVrn2'])\n",
    "                              \n",
    "    # Genotypes that have a positive VPPS show an acelleration of vernalisation under long day conditions.\n",
    "    # The molecular mechanium for this is uncertain so we attrubute it to VrnX\n",
    "    # Under Long Pp unvernalised varieties will reach VS HS sooner than BaseVrn1 expresion determines.\n",
    "    # BaseVrn1 expression at VS under long Pp unvernalised treatment (BaseVrn1AtVSHS_LN)\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['BaseVrn1AtVSHS_LN'] = min(inputs['VernalisationThreshold'],(data['VSHS_LN'] + data['EmergHS']) * data['BaseDVrn1'])\n",
    "    \n",
    "    # If we assume the expression of VrnX upregulates Vrn1 an equivelent amount then the amount of VrnX expression can be\n",
    "    # Estimated as the difference between Base Vrn1 at VS and the Vernalisation threshold.  The duration of expression is\n",
    "    # VSHS assuming VrnX is expressed from emergence until VS.\n",
    "    # Maximum rate of Vrnx Expression (MaxDVrnX)\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    data['MaxDVrnX'] = (inputs['VernalisationThreshold']-data['BaseVrn1AtVSHS_LN'])/(data['VSHS_LN'])\n",
    "    \n",
    "    return data     "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
