{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting C:\\Users\\Cflhxb\\AppData\\Local\\anaconda3\\Lib\\CAMP.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile C:\\Users\\Cflhxb\\AppData\\Local\\anaconda3\\Lib\\CAMP.py\n",
    "    \n",
    "import datetime as dt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math as math \n",
    "import MathsUtilities as MUte\n",
    "from scipy.optimize import curve_fit\n",
    "\n",
    "CampConstants = {\n",
    "    'k':-0.17,\n",
    "    'VIThreshold':1.0,\n",
    "    'TSThreshold':2.0,\n",
    "    'SlopeFLNvsTS':1.1,\n",
    "    'IntFLNvsTS':2.85,\n",
    "    'PpMax':16,\n",
    "    'PpMin':8,\n",
    "}\n",
    "\n",
    "def CalcdHS(Tt,Pp,HS,BasePhyllo,PhylloPpSens):\n",
    "    \"\"\"Calculate daily increase in Haun stage\n",
    "    Args:\n",
    "        Tt: Thermal time increment\n",
    "        Pp: The current photoperiod\n",
    "        HS: Current Haun Stage\n",
    "        BasePhyllo: Base phyllochron at 16h Pp between 3 and 7 HS\n",
    "        PhylloPpSens: Photo period sensitivity of phyllochron.  Relative increase in phyllochron as 8h Pp\n",
    "    Returns:\n",
    "        Daily increas in Haun Stage\n",
    "    \"\"\"\n",
    "    HS = max(0,HS)\n",
    "    StageFactor = np.interp(HS,[0,2,2.0001,7,7.0001,12],[0.75,0.75,1.0,1.0,1.4,1.4],False)\n",
    "    PpFactor = 1 + PhylloPpSens * np.interp(Pp,[0,8,12,20],[1,1,0,0],False)\n",
    "    Phyllochron = BasePhyllo*StageFactor*PpFactor\n",
    "    return Tt/Phyllochron\n",
    "\n",
    "def CalcPpResponse(Pp, Base, Max):\n",
    "    \"\"\" Calculate value of a photoperiod sensitive variable between Max and Baee values\n",
    "    Args:\n",
    "        Pp: Photoperiod\n",
    "        Base: value below 8h Pp\n",
    "        Max: value above 16h Pp\n",
    "        dPB: delta base phyllochrons\n",
    "    Returns:\n",
    "        variable value dependent on photoperiod and adjusted for dHS\n",
    "    \"\"\"\n",
    "    if Pp <= CampConstants['PpMin']:\n",
    "        return Base\n",
    "    if (Pp > CampConstants['PpMin']) and (Pp < CampConstants['PpMax']):\n",
    "        return (Base + (Max-Base) * (Pp-CampConstants['PpMin'])/(CampConstants['PpMax']-CampConstants['PpMin']))\n",
    "    if (Pp >= CampConstants['PpMax']):\n",
    "        return Max\n",
    "\n",
    "def CalcBaseUpRegVrn1(Tt, dHS, BaseDVrn1):\n",
    "    \"\"\" Calculate upregulation of base Vrn1\n",
    "    \n",
    "    Args:\n",
    "        Tt: Thermal time increment\n",
    "        dHS: Delta HaunStage\n",
    "        BaseDVrn1: coeffociennt for Base Vrn1 expression\n",
    "        \n",
    "    Returns:\n",
    "        delta BaseVrn1 representing the additional Vrn1 expression from base expression\n",
    "    \"\"\"\n",
    "    if Tt < 0: \n",
    "        BaseDVrn1 = 0\n",
    "    return BaseDVrn1 * dHS\n",
    "\n",
    "def CalcColdUpRegVrn1(Tt,dHS, MaxApDev1, k):\n",
    "    \"\"\" Upregulation of Vrn1 from cold.  Is additional to base vrn1\n",
    "        BaseDVrn1 in seperate calculation otherwise te same as Brown etal 2013\n",
    "    Args:\n",
    "        Tt: Thermal time increment\n",
    "        dHS: Delta Base Phyllochrons\n",
    "        MaxApDev1: coefficient for Maximum upregulation of Vrn1\n",
    "        k: The exponential coefficient determining temperature response\n",
    "        \n",
    "    Returns:\n",
    "        delta ColdVrn1 representing the additional Vrn1 expression from cold upregulation\n",
    "    \"\"\"\n",
    "    UdVrn1 = MaxApDev1 * np.exp(k*Tt)\n",
    "    if (Tt < 20):\n",
    "        return UdVrn1 * dHS\n",
    "    else:\n",
    "        return -5\n",
    "        \n",
    "def calcTSHS(FLN,IntFLNvsTSHS,SlopeFLNvsTS):\n",
    "    \"\"\"Haun stage timing of terminal spikelet \n",
    "       Inverts equation 5 from Brown etal 2013 FLN =  2.85 + 1.1*TSHS and converts it to base phyllochrons.\n",
    "       Note the intercept differs, was typeo on publication\n",
    "       Intercept has been made variable for as needs to be lower for some very fast varieties\n",
    "    Args:\n",
    "        FLN: The final leaf number\n",
    "        IntFLNvsTSHS: The intercept of the regression between FLN and TSHS\n",
    "    Returns:\n",
    "        Estimation of number of base phyllochrons to terminal spikelet\n",
    "    \"\"\"\n",
    "    return (FLN - IntFLNvsTSHS)/SlopeFLNvsTS\n",
    "\n",
    "def CAMPmodel(Out, Day, Tt, Pp, Params, Consts, TtEmerge):\n",
    "    \"\"\" The Cereal Anthesis Molecular Phenology model.\n",
    "        Based on the ideas presented in Brown etal 2014 (Annals of Botany)\n",
    "        Alterations made replacing Vrn4 notion with methalation of Vrn1, allowing Vrn3 to act before VI and changint the pattern for Vrn2 expression\n",
    "        Other alterations to implement working code base\n",
    "    Args:\n",
    "        Out: \"FLN\" or None.  If \"FLN\" will only return estimated FLN else will return full dataframe with daily state variable values\n",
    "        Day: List, 1:EndDay representing the timesteps in model run\n",
    "        Tt: List of same length as Day, representing daily temperature\n",
    "        Pp: List of same length as Day, representing daily photoperiod\n",
    "        Params: Dict with genotype parameters fitted by CalcCultivarVrnCoeffs() \n",
    "        Consts: Dict with crop specific constants,\n",
    "        TtEmerge: Thermaltime from sowing to emergence\n",
    "    \"\"\"\n",
    "    # Set up Data structure and initialise values\n",
    "    # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "    DF = pd.DataFrame(index = [0], columns = ['Day','Tt','Pp'])\n",
    "    IsGerminated = False\n",
    "    IsEmerged = False\n",
    "    IsMethalating = False\n",
    "    IsVernalised = False\n",
    "    IsReproductive = False\n",
    "    IsAtFlagLeaf = False\n",
    "    DF.loc[0,'Day'] = 0\n",
    "    DF.loc[0,'Tt'] = Tt[0]\n",
    "    DF.loc[0,'RelCold'] = 0\n",
    "    DF.loc[0,'Pp'] = Pp[0]\n",
    "    DF.loc[0,'RelPp'] = 0\n",
    "    DF.loc[0,'Stage'] = 'Germination'\n",
    "    DF.loc[0,'IsGerminated'] =True\n",
    "    DF.loc[0,'IsEmerged'] = False\n",
    "    DF.loc[0,'IsVernalised'] = False\n",
    "    DF.loc[0,'IsReproductive'] = False\n",
    "    DF.loc[0,'IsAtFlagLeaf'] = False\n",
    "    DF.loc[0,'HS'] = -TtEmerge/Params.BasePhyllo\n",
    "    DF.loc[0,'dHS'] = 0\n",
    "    DF.loc[0,'AccumTt'] = Tt[0]\n",
    "    DF.loc[0,'VrnB'] = 0\n",
    "    DF.loc[0,'dVrnB'] = 0\n",
    "    DF.loc[0,'MaxApDev'] = 0\n",
    "    DF.loc[0,'dMaxApDev'] = 0\n",
    "    DF.loc[0,'Cold'] = 0\n",
    "    DF.loc[0,'dCold'] = 0\n",
    "    DF.loc[0,'ApDev'] = 0\n",
    "    DF.loc[0,'MaxVrn2'] = 0\n",
    "    DF.loc[0,'Vrn2'] = 0.0\n",
    "    DF.loc[0,'AccumSPp'] = 0.0\n",
    "    #DF.loc[0,'Vrn2ef'] = 0.0\n",
    "    DF.loc[0,'Vrn1'] = 0.0\n",
    "    DF.loc[0,'dVrn1'] = 0\n",
    "    DF.loc[0,'Vrn3'] = 0.0\n",
    "    DF.loc[0,'dVrn3'] = 0\n",
    "    DF.loc[0,'VIHS'] = 0\n",
    "    DF.loc[0,'TSHS'] = 0\n",
    "    DF.loc[0,'FLN'] = Consts['IntFLNvsTS']\n",
    "    \n",
    "    d = 1\n",
    "    \n",
    "    # Model daily loop\n",
    "    # ^^^^^^^^^^^^^^^^\n",
    "    while (IsAtFlagLeaf == False) and (d < Day[-1]):\n",
    "        ColdYesterday = 0\n",
    "        ColdYesterday = DF.loc[d-1,'Cold']\n",
    "        DF.loc[d,:] = DF.loc[d-1,:] #Carry yesterdays values over to today\n",
    "        DF.loc[d,'Stage'] = np.nan\n",
    "            \n",
    "        DF.loc[d,'Day'] = d\n",
    "        # Set daily environment variables\n",
    "        DF.loc[d,'Tt'] = Tt[d]\n",
    "        DF.loc[d,'Pp'] = Pp[d]\n",
    "        #Zero set deltas\n",
    "        DF.loc[d,'dHS'] = 0\n",
    "        DF.loc[d,'dVrnB'] = 0\n",
    "        DF.loc[d,'dMaxApDev'] = 0\n",
    "        DF.loc[d,'dCold'] = 0\n",
    "        DF.loc[d,'Vrn2'] = 0\n",
    "        DF.loc[d,'dVrn1'] = 0\n",
    "        DF.loc[d,'dVrn3'] = 0\n",
    "        \n",
    "                \n",
    "        DF.loc[d,'AccumTt'] = DF.loc[d,'AccumTt'] + DF.loc[d,'Tt']\n",
    "        \n",
    "        PropnOfDay = 1.0\n",
    "        if (DF.loc[d,'AccumTt'] > TtEmerge) and (IsEmerged==False):\n",
    "            IsEmerged = True\n",
    "            DF.loc[d,'Stage'] = 'Emergence'\n",
    "            PropnOfDay = (DF.loc[d,'AccumTt'] - TtEmerge)/DF.loc[d,'Tt'] # Calculate fraction of emergence days Tt that is not used for emergence\n",
    "        \n",
    "        # Calculate daily Haun Stage changes\n",
    "        # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "        if IsEmerged==False: # Crop not yet emerged \n",
    "            EmergDurationFactor = 1\n",
    "            if (DF.loc[d,'AccumTt'] > 90):  # Calculate EmergenceDurationFactor to slow accumulation of HS if emergence is taking a long time.  This slows Vrn1 expression under slow emergence and strange responses to delayed sowing\n",
    "                EmergDurationFactor = np.exp(-0.015 * (DF.loc[d,'AccumTt']-90))\n",
    "            DF.loc[d,'dHS'] = DF.loc[d,'Tt']/Params.BasePhyllo  * EmergDurationFactor\n",
    "            DF.loc[d,'RelPp'] = 0\n",
    "            DF.loc[d,'Vrn2'] = 0\n",
    "            \n",
    "        else: # Crop emerged\n",
    "            # Calculate delta haun stage\n",
    "            DF.loc[d,'dHS'] = CalcdHS(DF.loc[d,'Tt'],DF.loc[d,'Pp'],DF.loc[d,'HS'],Params.BasePhyllo,Params.PhylloPpSens)\n",
    "            \n",
    "        #increment HS\n",
    "        DF.loc[d,'HS'] += DF.loc[d,'dHS']\n",
    "        \n",
    "        # Calculate Vrn gene expression\n",
    "        # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "        if (IsReproductive==False):\n",
    "            if (IsVernalised==False): #set as rates and factors for vegetative phase\n",
    "                DF.loc[d,'dVrnB'] = Params.BaseDVrnVeg * DF.loc[d,'dHS']\n",
    "                DF.loc[d,'dMaxApDev'] = Params.MaxApDevVeg * DF.loc[d,'dHS']\n",
    "                DF.loc[d,'PpVrn3Fact'] = Params.PpVrn3FactVeg\n",
    "            else:\n",
    "                 #set as rates and factors for early reproductive phase\n",
    "                DF.loc[d,'dVrnB'] = Params.BaseDVrnER * DF.loc[d,'dHS']\n",
    "                DF.loc[d,'dMaxApDev'] = Params.MaxApDevER * DF.loc[d,'dHS']\n",
    "                DF.loc[d,'PpVrn3Fact'] = Params.PpVrn3FactER\n",
    "            #Increment BaseVern and MaxVern\n",
    "            DF.loc[d,'VrnB'] += DF.loc[d,'dVrnB']\n",
    "            DF.loc[d,'MaxApDev'] = min(2,DF.loc[d,'MaxApDev'] + DF.loc[d,'dMaxApDev'])\n",
    "\n",
    "            # Calculate daily cold response\n",
    "            if (IsVernalised==False):\n",
    "                DF.loc[d,'RelCold'] = CalcColdUpRegVrn1(DF.loc[d,'Tt'],1,1,Consts['k'])\n",
    "                DF.loc[d,'dCold'] = DF.loc[d,'RelCold'] * DF.loc[d,'dVrnB'] * Params.ColdVrn1Fact\n",
    "                DF.loc[d,'Cold'] = max(0.0,DF.loc[d,'Cold'] + DF.loc[d,'dCold'])\n",
    "                if DF.loc[d,'Cold'] > Params.MethalationThreshold:\n",
    "                    DF.loc[d,'dVrn1'] = max(0,(DF.loc[d,'Cold'] - Params.MethalationThreshold)-(ColdYesterday - Params.MethalationThreshold))\n",
    "            # Increment Vrn 1\n",
    "            DF.loc[d,'Vrn1'] += DF.loc[d,'dVrn1']\n",
    "\n",
    "            # Calcualte expression of photoperiod sensitive genes\n",
    "            if (IsEmerged==True):  # Photoperiod sensitive genes only express after emergence\n",
    "                DF.loc[d,'RelPp'] = CalcPpResponse(DF.loc[d,'Pp'],0,1) #relative Pp, scaled between 0 at lower threshold and 1 at upper threshold\n",
    "                DF.loc[d,'AccumSPp'] += DF.loc[d,'dVrnB'] * (1-DF.loc[d,'RelPp'])  #accumulate short day exposure for reducing maxVrn2\n",
    "                DF.loc[d,'MaxVrn2'] = max(0,(Params.MaxVrn2 * DF.loc[d,'RelPp']) - DF.loc[d,'AccumSPp'])\n",
    "                DF.loc[d,'Vrn2'] = max(0,DF.loc[d,'MaxVrn2'] - (DF.loc[d,'VrnB'] + DF.loc[d,'Vrn1'])) \n",
    "                if (DF.loc[d,'Vrn2'] == 0): # express vrn3 relative to Pp if effective Vrn2 is down regulated to zero\n",
    "                    DF.loc[d,'dVrn3'] = (DF.loc[d,'PpVrn3Fact']-1) * DF.loc[d,'RelPp'] * DF.loc[d,'dVrnB']\n",
    "                #Increment Vrn3\n",
    "            DF.loc[d,'Vrn3'] += DF.loc[d,'dVrn3']\n",
    "\n",
    "            # Increment todays Vrn expression values using deltas just calculated\n",
    "            if (IsVernalised==False): #If not vernalised need to include Vrn2 and constrain to maxrate\n",
    "                DF.loc[d,'ApDev'] = min(DF.loc[d,'MaxApDev'],max(0,DF.loc[d,'VrnB'] + DF.loc[d,'Vrn1'] + DF.loc[d,'Vrn3'] - DF.loc[d,'MaxVrn2'])) \n",
    "            else:\n",
    "                DF.loc[d,'ApDev'] = DF.loc[d,'ApDev'] + DF.loc[d,'dVrnB'] + DF.loc[d,'dVrn3']\n",
    "\n",
    "        # Set Haun stage variables\n",
    "        # ^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "        if IsVernalised == False:\n",
    "            DF.loc[d,'VIHS'] = DF.loc[d,'HS']\n",
    "        if IsReproductive == False:\n",
    "            DF.loc[d,'TSHS'] = DF.loc[d,'HS']\n",
    "            DF.loc[d,'FLN'] = Consts['IntFLNvsTS'] + Consts['SlopeFLNvsTS'] * DF.loc[d,'TSHS']\n",
    "\n",
    "        # Finally determine phenological stage\n",
    "        # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "        # Vernalisation saturation occurs when Vrn1 > the vernalisation threshold and Vrn2 expression is zero\n",
    "        if (IsEmerged==True) and (DF.loc[d,'ApDev'] >= Consts['VIThreshold']) and (IsVernalised==False):\n",
    "            IsVernalised  = True\n",
    "            DF.loc[d,'Stage'] = 'Vern Sat'\n",
    "            #DF.loc[d,'Vrn1'] = 0\n",
    "        \n",
    "        if (DF.loc[d,'ApDev'] >=  Consts['TSThreshold']) and (IsReproductive == False):\n",
    "            IsReproductive = True;\n",
    "            DF.loc[d,'Stage'] = 'Term Spike'\n",
    "        \n",
    "        #Work out if Flag leaf has appeared.\n",
    "        if DF.loc[d,'HS'] >= DF.loc[d,'FLN']:\n",
    "            IsAtFlagLeaf = True    \n",
    "            DF.loc[d,'Stage'] = 'FlagLeaf'\n",
    "    \n",
    "        #Add states to dataframe\n",
    "        IsDayOfEmergence = False\n",
    "        \n",
    "        DF.loc[d,'IsGerminated'] = IsGerminated\n",
    "        DF.loc[d,'IsEmerged'] = IsEmerged\n",
    "        DF.loc[d,'IsVernalised'] = IsVernalised\n",
    "        DF.loc[d,'IsReproductive'] = IsReproductive\n",
    "        DF.loc[d,'IsAtFlagLeaf'] = IsAtFlagLeaf\n",
    "    \n",
    "        # Increment day\n",
    "        d += 1\n",
    "        \n",
    "    if (Out == 'FLN') and (IsAtFlagLeaf == True):\n",
    "        return DF.iloc[-1,:]['FLN']\n",
    "    else:\n",
    "        return DF\n",
    "    \n",
    "def deriveVrnParams(c):\n",
    "    data = pd.Series(dtype=np.float64)\n",
    "    # Calculate the accumulated base phyllochrons at terminal spikelet for each treatment\n",
    "    for pv in ['SN','LV','SV','LN']:\n",
    "        data['TS_'+pv] = calcTSHS(c['FLN_'+pv],CampConstants['IntFLNvsTS'],CampConstants['SlopeFLNvsTS'])\n",
    "    # Base Phyllochron\n",
    "    data['BasePhyllo'] = c['BasePhyllo']\n",
    "    # Phyllochron photoperiod sensitivity.  #This is not used in calculations but added so can be checked in outputs\n",
    "    data['PhylloPpSens'] = c['PhylloPpSens']\n",
    "    # Base Phyllochron duration of the Emergence Phase\n",
    "    data['EmergDurat'] = c['TtEmerge']/data['BasePhyllo']\n",
    "    # Base Phyllochron duration of vernalisation treatment\n",
    "    data['VernTreatDurat'] = (c['VrnTreatDuration']*c['VrnTreatTemp'])/data['BasePhyllo']\n",
    "    # Accumulated Base Phyllochrons when vernalisation treatment ended \n",
    "    data['EndVernTreat'] =  -data['EmergDurat'] + data['VernTreatDurat']\n",
    "    # The minimum accumulated haunstage at which Vern saturation can occur\n",
    "    data['MinVI'] = 1.1\n",
    "    # The minimum haun stage duration from vern saturation to terminal spikelet\n",
    "    data['MinVI->TS'] = min(3,data['TS_LV']-data['MinVI'])  #From Lincoln controlled environment data\n",
    "    \n",
    "    # Calculate the accumulated base phyllochrons at vernalisation saturation for each treatment\n",
    "    data['VI_LV'] = data['TS_LV'] - data['MinVI->TS']   \n",
    "    data['VI_LN'] = max(data['TS_LN'] - data['MinVI->TS'],data['VI_LV']) #Constrained so not earlier than the LV treatment\n",
    "    data['VI_SV'] = min(data['VI_LV'],data['TS_SV']-data['MinVI->TS']) # Assume happens at the same time as VI_LV but not sooner than TS and MinVI->TS would allow \n",
    "    data['VI->TS_SV'] = data['TS_SV'] - data['VI_SV']\n",
    "    data['VI_SN'] = max(data['TS_SN'] - data['VI->TS_SV'],data['VI_LV']) #Constrained so not earlier than the LV treatment\n",
    "    \n",
    "    # Calculate base and maximum rates and Pp sensitivities\n",
    "    # Base Vrn delta during vegetative phase.  Assuming base Vrn expression starts at sowing and reaches 1 at VI for the SN treatment where no cold or Vrn3 (short Pp) to upregulate \n",
    "    data['BaseDVrnVeg'] = 1/(data['VI_SN']+data['EmergDurat'])\n",
    "    # The fastest rate that Vrn can accumulate to reach saturation (VI).\n",
    "    data['MaxApDevVeg'] = 1/(data['VI_SV']+data['EmergDurat'])\n",
    "    # Base Vrn delta during early reproductive phase.  Assuming Vrn expression increases by 1 between VI and TS and proceeds at a base rate where not Vrn3 upregulation (short PP)\n",
    "    data['BaseDVrnER'] = 1/(data['TS_SN']-data['VI_SN'])\n",
    "    # The maximum Vrn delta during the early reproductive phase.  Assuming Vrn increases by 1 bewtween VI and TS and proceeds at the maximum rate under long photoperiod\n",
    "    data['MaxApDevER'] = 1/data['MinVI->TS']\n",
    "    # The relative increase in delta Vrn cuased by Vrn3 expression under long photoperiod during early reproductive phase\n",
    "    data['PpVrn3FactER'] = data['MaxApDevER']/data['BaseDVrnER']\n",
    "    # The relative increase in delta Vrn1 caused by Vrn3 expression under long Pp during vegetative phase.  Same as PpVrn3FactER unless VI_LN is small\n",
    "    data['PpVrn3FactVeg'] = (1/data['VI_LN'])/data['BaseDVrnVeg'] \n",
    "    data['PpVrn3FactVeg'] = max(data['PpVrn3FactER'],data['PpVrn3FactVeg']) \n",
    "    \n",
    "    # Calculate Maximum Vrn2 expression under long photoperiods\n",
    "    # The rate of Vrn expression before VI under long Pp when Vrn2 = 0 is the product of Vrn3 and base Vrn\n",
    "    data['dvrnxVrn3'] = data['BaseDVrnVeg'] * data['PpVrn3FactVeg']\n",
    "    # Therefore the haun stage duration when effective baseVrn x Vrn3 is up regulating during the vegetative phase under long photoperiod without vernalisation is a funciton of the VI target (1) and the rate\n",
    "    data['vrnxVrn3Durat'] = 1/data['dvrnxVrn3']\n",
    "    # The accumulated Haun stages when Vrn2 expression is ended and effective baseVrn x Vrn3 expression starts under long Pp without vernalisation is vrn1xVrn3Durat prior to VI\n",
    "    data['endVrn2_LN'] = max(0,data['VI_LN'] - data['vrnxVrn3Durat']) \n",
    "    # The Vrn2 Expression that must be matched by Vrn1 before effective Vrn3 expression starts.  \n",
    "    data['MaxVrn2'] = (data['endVrn2_LN']+data['EmergDurat']) * data['BaseDVrnVeg']\n",
    "    \n",
    "    # Calculate cold upregulation of Vrn1\n",
    "    # The amount of Vrn expression at base rate in the LV treatment between sowing and when Vrn2 is suppressed \n",
    "    data['baseVrnVeg_LV'] = (data['EmergDurat']+data['VI_LV']) * data['BaseDVrnVeg']\n",
    "    # The amount of persistant (methalated) Vrn1 upregulated due to cold at the end of the vernalisation treatment for LV\n",
    "    # Is the VI threshold (1) plus the maximum Vrn2 less base vrn expression up to VI and Vrn3 upregulated expression between end of Vrn2 expression and VI\n",
    "    data['coldVrn1_LV'] = CampConstants['VIThreshold'] + data['MaxVrn2'] - data['baseVrnVeg_LV'] \n",
    "    # Cold threshold required before Vrn1 starts methalating.  Based on Brooking and Jamieson data the lag duration is the same length as the duration of response up to full vernalisation\n",
    "    # Therefore we assume the methalation threshold must be the same size as the amount of vrn1 that is required to give full persistend vernalisation response.\n",
    "    data['MethalationThreshold'] = data['coldVrn1_LV']\n",
    "    # Haun stage duration of vernalisation, may be less than treatment duration if treatment goes past VI\n",
    "    data['vernDurat_LV'] =  min(data['VernTreatDurat'],data['VI_LV']+data['EmergDurat'])\n",
    "    # The rate of Vrn1 expression under cold treatment calculated from the amopunt of cold vrn1 up regulation apparante plus the methalation threshold\n",
    "    data['coldDVrn1_LV'] = (data['coldVrn1_LV']+data['MethalationThreshold'])/data['vernDurat_LV'] \n",
    "    # The upregulation of Vrn1 expression above base rate at 0oC\n",
    "    data['coldDVrn1Max'] = data['coldDVrn1_LV']/ np.exp(c['k'] * c['VrnTreatTemp']) \n",
    "    # The relative increase in delta Vrn1 caused by cold upregulation of Vrn1\n",
    "    data['ColdVrn1Fact'] = data['coldDVrn1Max'] /data['BaseDVrnVeg']\n",
    "                                 \n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
